File: tests/test_moe_multi_token_fp4.py
"""Small-batch FP4 MoE kernel: agreement with the single-token kernel.

The point of moe_multi_token_fp4_forward is to serve a speculative-verify batch
(k+1 tokens) without falling into the prefill grouped path. It is only useful if
it agrees with the one-token kernel that decode already uses: a batch kernel
that shifts logits changes which draft tokens get accepted, which is how the
existing FP4_DIRECT_GROUPED path fails (max logit diff up to 1.48 on the real
model).

So the reference here is moe_single_token_fp4_forward applied one token at a
time. Both kernels use the same dp4a math over the same fp4 unpack and the same
per-row int8 activation scale, so they should agree to within float32 summation
order.

Bit-exactness is the wrong bar: both kernels accumulate a token's top-k
contributions with atomicAdd, so neither is deterministic. Measured on this box,
the reference kernel disagrees with ITSELF by up to ~9e-7 relative across
repeated runs of the same inputs. The tolerance below is set from that floor, and
`test_reference_is_not_bit_exact` pins the reasoning in place so a future reader
does not tighten it to zero and get a phantom failure.
"""
from __future__ import annotations

import os

import pytest
import torch

cuda_kernel = pytest.importorskip("cuda_kernel")
pytestmark = pytest.mark.skipif(not torch.cuda.is_available(), reason="requires CUDA")

DIM = 256
INTER_DIM = 128
N_EXPERTS = 8
TOPK = 3
SWIGLU_LIMIT = 7.0
# Fallback tolerance for the shapes where the reference's own spread is measured
# to be tiny. Tests that can measure the floor prefer the measured value: the
# spread grows with how many routes are summed per token (~9e-7 relative at
# topk=3, ~8e-4 at topk=8), so no single constant fits every shape.
RTOL = 1e-4


def _random_fp4_weights(n_experts: int, rows: int, k: int, device, seed: int):
    """Random packed FP4 blocks: two 4-bit codes per byte, one e8m0 scale/32."""
    gen = torch.Generator(device=device).manual_seed(seed)
    q = torch.randint(0, 256, (n_experts, rows, k // 2), dtype=torch.uint8,
                      device=device, generator=gen)
    # Keep exponents in a narrow band so the reference and the kernel are
    # compared on values that neither overflow nor collapse to zero.
    s = torch.randint(120, 136, (n_experts, rows, k // 32), dtype=torch.uint8,
                      device=device, generator=gen)
    return q.contiguous(), s.contiguous()


def _build_routes(indices: torch.Tensor, weights: torch.Tensor, n_local: int):
    """Compact [T, topk] routing into the kernel's slot/pair layout.

    Only experts the batch actually touches get a slot, which is what keeps the
    launch proportional to the batch instead of to n_local_experts.
    """
    tokens, topk = indices.shape
    flat_e = indices.reshape(-1)
    flat_t = torch.arange(tokens, device=indices.device).repeat_interleave(topk)
    flat_w = weights.reshape(-1).to(torch.float32)
    keep = (flat_e >= 0) & (flat_e < n_local)
    flat_e, flat_t, flat_w = flat_e[keep], flat_t[keep], flat_w[keep]
    order = torch.argsort(flat_e, stable=True)
    flat_e, flat_t, flat_w = flat_e[order], flat_t[order], flat_w[order]
    slot_expert, counts = torch.unique_consecutive(flat_e, return_counts=True)
    slot_starts = torch.zeros(slot_expert.numel() + 1, dtype=torch.int32,
                              device=indices.device)
    slot_starts[1:] = counts.cumsum(0).to(torch.int32)
    return (slot_expert.to(torch.int32).contiguous(),
            slot_starts.contiguous(),
            flat_t.to(torch.int32).contiguous(),
            flat_w.contiguous())


def _reference_per_token(x, indices, weights, w, swiglu_limit):
    """Run the production single-token kernel once per row."""
    outs = []
    for t in range(x.size(0)):
        outs.append(cuda_kernel.moe_single_token_fp4_forward(
            x[t:t + 1].contiguous(),
            indices[t].contiguous().to(torch.int64),
            weights[t].contiguous().to(torch.float32),
            *w, 0, swiglu_limit,
        ))
    return torch.cat(outs, dim=0)


def _rel_spread(runs) -> float:
    return max(float(((runs[0] - r).abs() / runs[0].abs().clamp_min(1e-9)).max())
               for r in runs[1:])


def _assert_within_reference_noise(got, want, ref_runs, label: str):
    """Compare against the reference's own spread, floored at a fixed epsilon.

    Historically both kernels combined a token's routes with atomicAdd, so the
    reference disagreed with itself and that spread was the natural bound. The
    reduction is deterministic now, so the spread term is normally 0 and RTOL
    carries the bound; the max() is kept so the assertion still adapts if the
    atomic path is selected via DEEPSEEK_MOE_DETERMINISTIC_REDUCE=0.
    """
    floor = max(_rel_spread(ref_runs), RTOL)
    diff = (got - want).abs()
    rel = float((diff / want.abs().clamp_min(1e-9)).max())
    assert rel <= floor * 4.0, (
        f"{label}: cross-kernel max_rel={rel:.3e} exceeds 4x the reference's own "
        f"run-to-run spread ({floor:.3e}); the batching changed the arithmetic")


def _run_pair(tokens: int, seed: int, topk: int = TOPK):
    device = torch.device("cuda")
    gen = torch.Generator(device=device).manual_seed(seed)
    x = torch.randn(tokens, DIM, dtype=torch.bfloat16, device=device, generator=gen)
    w1q, w1s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, seed + 1)
    w3q, w3s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, seed + 2)
    w2q, w2s = _random_fp4_weights(N_EXPERTS, DIM, INTER_DIM, device, seed + 3)
    w = (w1q, w1s, w2q, w2s, w3q, w3s)

    indices = torch.stack([
        torch.randperm(N_EXPERTS, device=device, generator=gen)[:topk]
        for _ in range(tokens)
    ]).to(torch.int64)
    weights = torch.rand(tokens, topk, dtype=torch.float32, device=device, generator=gen) + 0.1

    slot_expert, slot_starts, slot_tokens, pair_weights = _build_routes(
        indices, weights, N_EXPERTS)
    got = cuda_kernel.moe_multi_token_fp4_forward(
        x, slot_expert, slot_starts, slot_tokens, pair_weights, *w, SWIGLU_LIMIT)
    # Several reference runs: the first is the value to match, the spread across
    # them is the noise floor that match is judged against.
    ref_runs = [_reference_per_token(x, indices, weights, w, SWIGLU_LIMIT)
                for _ in range(4)]
    return got, ref_runs[0], ref_runs


@pytest.mark.parametrize("tokens", [1, 2, 3, 6])
def test_matches_single_token_kernel(tokens):
    """A verify batch must produce what the same tokens produce one at a time."""
    got, want, ref_runs = _run_pair(tokens, seed=1000 + tokens)
    assert got.shape == want.shape == (tokens, DIM)
    _assert_within_reference_noise(got, want, ref_runs, f"tokens={tokens}")


def test_reference_is_not_bit_exact():
    """Guards the tolerance above: the reference kernel is itself nondeterministic.

    Both kernels atomicAdd a token's top-k contributions, so summation order
    varies between launches. If this ever starts passing at rtol=0, the
    tolerance in the other tests can be tightened -- until then, demanding
    bit-exactness would be testing the GPU scheduler, not the kernel.
    """
    device = torch.device("cuda")
    gen = torch.Generator(device=device).manual_seed(4242)
    x = torch.randn(2, DIM, dtype=torch.bfloat16, device=device, generator=gen)
    w1q, w1s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, 11)
    w3q, w3s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, 12)
    w2q, w2s = _random_fp4_weights(N_EXPERTS, DIM, INTER_DIM, device, 13)
    w = (w1q, w1s, w2q, w2s, w3q, w3s)
    indices = torch.stack([
        torch.randperm(N_EXPERTS, device=device, generator=gen)[:TOPK] for _ in range(2)
    ]).to(torch.int64)
    weights = torch.rand(2, TOPK, dtype=torch.float32, device=device, generator=gen) + 0.1
    runs = [_reference_per_token(x, indices, weights, w, SWIGLU_LIMIT) for _ in range(5)]
    spread = max(float((runs[0] - r).abs().max()) for r in runs[1:])
    rel = max(float(((runs[0] - r).abs() / runs[0].abs().clamp_min(1e-9)).max())
              for r in runs[1:])
    # Whatever the spread is, it must stay well inside the tolerance the
    # cross-kernel tests use, or those tests are not measuring the kernel.
    assert rel < RTOL, f"reference spread {rel:.2e} exceeds test tolerance {RTOL:.0e}"
    print(f"reference self-disagreement: max_abs={spread:.3g} max_rel={rel:.2e}")


def test_tokens_above_register_tile():
    """More tokens per expert than the register tile takes a second pass."""
    # topk == N_EXPERTS puts every token on every expert, so with 10 tokens each
    # slot holds 10 -- past the kernel's 8-wide tile. It also sums 8 routes per
    # token, which is where the atomicAdd noise floor is widest (~8e-4).
    got, want, ref_runs = _run_pair(10, seed=77, topk=N_EXPERTS)
    _assert_within_reference_noise(got, want, ref_runs, "tokens=10 topk=8")


def test_rows_for_non_local_experts_are_zero():
    """Experts outside this rank's range contribute nothing.

    Under TP each rank holds a slice of the experts and drops routes outside it.
    The caller filters those out, so an empty route list must yield zeros rather
    than reading out of bounds.
    """
    device = torch.device("cuda")
    x = torch.randn(3, DIM, dtype=torch.bfloat16, device=device)
    w1q, w1s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, 5)
    w3q, w3s = _random_fp4_weights(N_EXPERTS, INTER_DIM, DIM, device, 6)
    w2q, w2s = _random_fp4_weights(N_EXPERTS, DIM, INTER_DIM, device, 7)
    empty_i32 = torch.empty(0, dtype=torch.int32, device=device)
    y = cuda_kernel.moe_multi_token_fp4_forward(
        x,
        empty_i32,
        torch.zeros(1, dtype=torch.int32, device=device),
        empty_i32,
        torch.empty(0, dtype=torch.float32, device=device),
        w1q, w1s, w2q, w2s, w3q, w3s, SWIGLU_LIMIT,
    )
    assert torch.count_nonzero(y) == 0


class _FakeCPUBackend:
    """Minimal stand-in for the CPU MoE backend's staging interface.

    _forward_multi_token_fp4 only needs the FP4 arena and a layer index; the
    real backend brings a pinned host arena and a CPU expert pool that are
    irrelevant to the route bookkeeping under test here.
    """

    def __init__(self, arena):
        self._arena = arena
        self.layer_idx = 0
        self._swiglu_limit = SWIGLU_LIMIT

    def get_fp4_arena(self):
        return self._arena

    def _apply_topk_limit(self, ids, w):
        return ids, w


def _backend_with_arena(device, n_local, seed):
    """A GPUPrefillMoEBackend wired to host-pinned FP4 weights."""
    from src.components.moe.gpu_prefill_backend import GPUPrefillMoEBackend

    w1q, w1s = _random_fp4_weights(n_local, INTER_DIM, DIM, device, seed + 1)
    w3q, w3s = _random_fp4_weights(n_local, INTER_DIM, DIM, device, seed + 2)
    w2q, w2s = _random_fp4_weights(n_local, DIM, INTER_DIM, device, seed + 3)
    gpu_w = (w1q, w1s, w2q, w2s, w3q, w3s)
    arena = tuple(t.to("cpu") for t in gpu_w)
    backend = GPUPrefillMoEBackend(
        _FakeCPUBackend(arena), dim=DIM, num_experts=n_local,
        experts_start_idx=0, experts_end_idx=n_local,
    )
    backend.multi_token_fp4_enabled = True
    return backend, gpu_w


def test_backend_route_layout_handles_non_contiguous_indices():
    """The decode-active call site does not make indices contiguous.

    runtime.py's _should_use_gpu_decode_active_moe branch hands the gate's
    indices/weights straight to forward(), unlike the prefill branch which calls
    .contiguous() first. A transposed view reshapes in expert-major order, so
    pairing route i with token i // topk silently mismatches routes to tokens --
    which showed up on the real model as logits differing by up to 3.65 with
    only 50-67% argmax agreement, while the unit tests (building their own
    contiguous routes) passed.
    """
    device = torch.device("cuda")
    n_local, tokens, topk = N_EXPERTS, 3, TOPK
    backend, gpu_w = _backend_with_arena(device, n_local, seed=900)
    gen = torch.Generator(device=device).manual_seed(901)
    x = torch.randn(tokens, DIM, dtype=torch.bfloat16, device=device, generator=gen)

    # Build routing whose natural layout is [topk, tokens], then transpose: same
    # values as a contiguous [tokens, topk] tensor, different memory order.
    idx_t = torch.stack([
        torch.randperm(n_local, device=device, generator=gen)[:tokens] for _ in range(topk)
    ]).to(torch.int64)                       # [topk, tokens]
    w_t = torch.rand(topk, tokens, dtype=torch.float32, device=device, generator=gen) + 0.1
    indices_nc, weights_nc = idx_t.t(), w_t.t()   # [tokens, topk] views
    assert not indices_nc.is_contiguous()

    got = backend._forward_multi_token_fp4(
        x, indices_nc, weights_nc, SWIGLU_LIMIT)
    assert got is not None, "FP4 arena was available; the path should have run"
    want = _reference_per_token(
        x, indices_nc.contiguous(), weights_nc.contiguous(), gpu_w, SWIGLU_LIMIT)
    ref_runs = [_reference_per_token(
        x, indices_nc.contiguous(), weights_nc.contiguous(), gpu_w, SWIGLU_LIMIT)
        for _ in range(4)]
    _assert_within_reference_noise(got, want, ref_runs, "non-contiguous indices")


def test_backend_route_layout_matches_reference_contiguous():
    """Same check with a contiguous batch, so the two cases can be compared."""
    device = torch.device("cuda")
    n_local, tokens, topk = N_EXPERTS, 6, TOPK
    backend, gpu_w = _backend_with_arena(device, n_local, seed=910)
    gen = torch.Generator(device=device).manual_seed(911)
    x = torch.randn(tokens, DIM, dtype=torch.bfloat16, device=device, generator=gen)
    indices = torch.stack([
        torch.randperm(n_local, device=device, generator=gen)[:topk] for _ in range(tokens)
    ]).to(torch.int64)
    weights = torch.rand(tokens, topk, dtype=torch.float32, device=device, generator=gen) + 0.1

    got = backend._forward_multi_token_fp4(x, indices, weights, SWIGLU_LIMIT)
    assert got is not None
    ref_runs = [_reference_per_token(x, indices, weights, gpu_w, SWIGLU_LIMIT)
                for _ in range(4)]
    _assert_within_reference_noise(got, ref_runs[0], ref_runs, "contiguous indices")


# Production shapes: the TP4 rank slice of DeepSeek-V4-Flash (dim 4096, moe_inter_dim
# 2048, n_activated_experts 6, 256/4 = 64 local experts). The small shapes above
# exercise the code paths; this one pins the tolerance at the arithmetic width
# the model actually runs, where more routes per token widen the atomicAdd floor.
REAL_DIM, REAL_INTER, REAL_NLOC, REAL_TOPK = 4096, 2048, 64, 6


@pytest.mark.parametrize("tokens", [2, 3, 6])
def test_agreement_at_production_shapes(tokens):
    """Agreement measured on the layer output, not on logits.

    The real-model probes compared logits after 43 layers and lm_head, where a
    per-layer relative error is amplified and mixed with attention's own
    batch-vs-sequential differences -- so a 0.29 logit delta there could not be
    attributed. Comparing the MoE layer output directly makes the number the
    kernel's own, and it lands at the reference's atomicAdd noise floor.
    """
    device = torch.device("cuda")
    seed = 4000 + tokens
    gen = torch.Generator(device=device).manual_seed(seed)
    x = torch.randn(tokens, REAL_DIM, dtype=torch.bfloat16, device=device, generator=gen)
    w1q, w1s = _random_fp4_weights(REAL_NLOC, REAL_INTER, REAL_DIM, device, seed + 1)
    w3q, w3s = _random_fp4_weights(REAL_NLOC, REAL_INTER, REAL_DIM, device, seed + 2)
    w2q, w2s = _random_fp4_weights(REAL_NLOC, REAL_DIM, REAL_INTER, device, seed + 3)
    w = (w1q, w1s, w2q, w2s, w3q, w3s)
    indices = torch.stack([
        torch.randperm(REAL_NLOC, device=device, generator=gen)[:REAL_TOPK]
        for _ in range(tokens)
    ]).to(torch.int64)
    weights = torch.rand(tokens, REAL_TOPK, dtype=torch.float32, device=device,
                         generator=gen) + 0.1

    routes = _build_routes(indices, weights, REAL_NLOC)
    got = cuda_kernel.moe_multi_token_fp4_forward(x, *routes, *w, SWIGLU_LIMIT)
    ref_runs = [_reference_per_token(x, indices, weights, w, SWIGLU_LIMIT)
                for _ in range(4)]
    assert got.shape == (tokens, REAL_DIM)
    # Both kernels are now deterministic (DEEPSEEK_MOE_DETERMINISTIC_REDUCE
    # defaults on), so run-to-run spread is exactly zero and cannot serve as a
    # noise floor -- it would collapse the bound onto RTOL and fail a
    # disagreement that has always been there. Assert reproducibility separately
    # from magnitude, and bound the magnitude against int8 hidden-state
    # quantization, which is what actually sets it: the reference requantizes per
    # token while the kernel requantizes per (slot, token) pair.
    ours = [cuda_kernel.moe_multi_token_fp4_forward(x, *routes, *w, SWIGLU_LIMIT)
            for _ in range(3)]
    if os.environ.get("DEEPSEEK_MOE_DETERMINISTIC_REDUCE", "1") != "0":
        assert _rel_spread([got] + ours) == 0.0, (
            "kernel is not run-to-run reproducible")
        assert _rel_spread(ref_runs) == 0.0, (
            "reference is not run-to-run reproducible")
    rel = float(((got - ref_runs[0]).abs()
                 / ref_runs[0].abs().clamp_min(1e-9)).max())
    assert rel <= 8e-3, (
        f"tokens={tokens} at production shapes: max_rel={rel:.3e} exceeds the "
        f"int8 requantization bound; the batching changed the arithmetic")


def test_one_token_per_slot_is_the_single_token_computation():
    """Disjoint routing per token means every slot holds one token.

    With one token per slot the multi-token kernel does not enter its tile loop
    and performs the same arithmetic as the single-token kernel, so agreement
    here isolates the slot bookkeeping from the multi-token accumulation. Swept
    over many routings because the real-model bs=2 probe disagreed on 1 of 3
    draws, which a single seed would have missed.
    """
    device = torch.device("cuda")
    n_experts, tokens, topk = 16, 2, 3
    worst, worst_seed = 0.0, None
    for seed in range(24):
        gen = torch.Generator(device=device).manual_seed(seed)
        x = torch.randn(tokens, DIM, dtype=torch.bfloat16, device=device, generator=gen)
        w1q, w1s = _random_fp4_weights(n_experts, INTER_DIM, DIM, device, seed + 1)
        w3q, w3s = _random_fp4_weights(n_experts, INTER_DIM, DIM, device, seed + 2)
        w2q, w2s = _random_fp4_weights(n_experts, DIM, INTER_DIM, device, seed + 3)
        w = (w1q, w1s, w2q, w2s, w3q, w3s)
        perm = torch.randperm(n_experts, device=device, generator=gen)
        indices = torch.stack([perm[:topk], perm[topk:2 * topk]]).to(torch.int64)
        weights = torch.rand(tokens, topk, dtype=torch.float32, device=device,
                             generator=gen) + 0.1
        routes = _build_routes(indices, weights, n_experts)
        got = cuda_kernel.moe_multi_token_fp4_forward(x, *routes, *w, SWIGLU_LIMIT)
        ref = _reference_per_token(x, indices, weights, w, SWIGLU_LIMIT)
        rel = float(((got - ref).abs() / ref.abs().clamp_min(1e-9)).max())
        if rel > worst:
            worst, worst_seed = rel, seed
    assert worst < 1e-3, (
        f"one token per slot should match the single-token kernel to the atomicAdd "
        f"floor; worst max_rel={worst:.3e} at seed={worst_seed}")

File: cpp_engine/tests/test_moe_multi_token_fp4.cpp
// Active-slot small-batch FP4 MoE agreement with repeated single-token calls.

#include "cuda_ops.hpp"

#include <cuda_runtime.h>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

void check_cuda(cudaError_t err, const char* what) {
    if (err != cudaSuccess) {
        throw std::runtime_error(std::string(what) + ": " + cudaGetErrorString(err));
    }
}

template <typename T>
struct DeviceBuffer {
    T* ptr = nullptr;
    DeviceBuffer() = default;
    explicit DeviceBuffer(size_t count) {
        if (count != 0) check_cuda(cudaMalloc(&ptr, count * sizeof(T)), "cudaMalloc");
    }
    DeviceBuffer(const DeviceBuffer&) = delete;
    DeviceBuffer& operator=(const DeviceBuffer&) = delete;
    ~DeviceBuffer() { if (ptr != nullptr) cudaFree(ptr); }
};

template <typename T>
void copy_to_device(DeviceBuffer<T>& dst, const std::vector<T>& src) {
    if (!src.empty()) {
        check_cuda(cudaMemcpy(dst.ptr, src.data(), src.size() * sizeof(T),
                              cudaMemcpyHostToDevice), "copy to device");
    }
}

struct RouteLayout {
    std::vector<int32_t> experts;
    std::vector<int32_t> starts;
    std::vector<int32_t> tokens;
    std::vector<float> weights;
};

RouteLayout build_routes(const std::vector<int64_t>& indices,
                         const std::vector<float>& weights,
                         int tokens, int topk, int n_local) {
    RouteLayout out;
    out.starts.push_back(0);
    for (int expert = 0; expert < n_local; ++expert) {
        bool active = false;
        for (int token = 0; token < tokens; ++token) {
            for (int k = 0; k < topk; ++k) {
                const int idx = token * topk + k;
                if (indices[idx] != expert) continue;
                if (!active) {
                    out.experts.push_back(expert);
                    active = true;
                }
                out.tokens.push_back(token);
                out.weights.push_back(weights[idx]);
            }
        }
        if (active) out.starts.push_back(static_cast<int32_t>(out.tokens.size()));
    }
    return out;
}

struct CaseBuffers {
    DeviceBuffer<float> x, y, ref, weights, pair_weights;
    DeviceBuffer<int64_t> indices;
    DeviceBuffer<int32_t> slot_expert, slot_starts, slot_tokens;
    DeviceBuffer<uint8_t> w1q, w1s, w2q, w2s, w3q, w3s;
    DeviceBuffer<int8_t> x_q, hidden_q;
    DeviceBuffer<float> x_scale, gate, up, hidden_scale, partials;

    CaseBuffers(int tokens, int topk, int slots, int pairs, int n_local,
                int dim, int inter)
        : x(static_cast<size_t>(tokens) * dim),
          y(static_cast<size_t>(tokens) * dim),
          ref(static_cast<size_t>(tokens) * dim),
          weights(static_cast<size_t>(tokens) * topk),
          pair_weights(pairs),
          indices(static_cast<size_t>(tokens) * topk),
          slot_expert(slots), slot_starts(static_cast<size_t>(slots) + 1),
          slot_tokens(pairs),
          w1q(static_cast<size_t>(n_local) * inter * (dim / 2)),
          w1s(static_cast<size_t>(n_local) * inter * (dim / 32)),
          w2q(static_cast<size_t>(n_local) * dim * (inter / 2)),
          w2s(static_cast<size_t>(n_local) * dim * (inter / 32)),
          w3q(static_cast<size_t>(n_local) * inter * (dim / 2)),
          w3s(static_cast<size_t>(n_local) * inter * (dim / 32)),
          x_q(static_cast<size_t>(tokens) * dim), x_scale(tokens),
          gate(static_cast<size_t>(pairs) * inter),
          up(static_cast<size_t>(pairs) * inter),
          hidden_q(static_cast<size_t>(pairs) * inter), hidden_scale(pairs),
          partials(static_cast<size_t>(pairs) * dim) {}
};

void fill_bytes(uint8_t* dst, size_t count, std::mt19937& rng, bool scale) {
    std::vector<uint8_t> host(count);
    if (scale) {
        std::uniform_int_distribution<int> dist(120, 135);
        for (auto& v : host) v = static_cast<uint8_t>(dist(rng));
    } else {
        for (auto& v : host) v = static_cast<uint8_t>(rng() & 0xff);
    }
    check_cuda(cudaMemcpy(dst, host.data(), host.size(), cudaMemcpyHostToDevice),
               "copy random bytes");
}

void compare(const std::vector<float>& got, const std::vector<float>& want,
             float rel_limit, const std::string& label) {
    float max_abs = 0.0f;
    float max_rel = 0.0f;
    for (size_t i = 0; i < got.size(); ++i) {
        const float diff = std::fabs(got[i] - want[i]);
        max_abs = std::max(max_abs, diff);
        max_rel = std::max(max_rel, diff / std::max(std::fabs(want[i]), 1.0e-9f));
    }
    std::cout << label << " max_abs=" << max_abs << " max_rel=" << max_rel << "\n";
    if (max_rel > rel_limit) {
        throw std::runtime_error(label + ": relative error exceeds bound");
    }
}

void run_case(int tokens, int topk, int n_local, int dim, int inter,
              int seed, float rel_limit, bool all_experts) {
    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> xdist(-1.0f, 1.0f);
    std::uniform_real_distribution<float> wdist(0.1f, 1.1f);
    std::vector<float> x(static_cast<size_t>(tokens) * dim);
    for (auto& v : x) v = xdist(rng);
    std::vector<int64_t> indices(static_cast<size_t>(tokens) * topk);
    std::vector<float> weights(indices.size());
    for (int token = 0; token < tokens; ++token) {
        std::vector<int> experts(n_local);
        for (int e = 0; e < n_local; ++e) experts[e] = e;
        if (!all_experts) std::shuffle(experts.begin(), experts.end(), rng);
        for (int k = 0; k < topk; ++k) {
            indices[token * topk + k] = experts[k];
            weights[token * topk + k] = wdist(rng);
        }
    }
    const RouteLayout routes = build_routes(indices, weights, tokens, topk, n_local);
    const int slots = static_cast<int>(routes.experts.size());
    const int pairs = static_cast<int>(routes.tokens.size());
    CaseBuffers b(tokens, topk, slots, pairs, n_local, dim, inter);
    copy_to_device(b.x, x);
    copy_to_device(b.indices, indices);
    copy_to_device(b.weights, weights);
    copy_to_device(b.slot_expert, routes.experts);
    copy_to_device(b.slot_starts, routes.starts);
    copy_to_device(b.slot_tokens, routes.tokens);
    copy_to_device(b.pair_weights, routes.weights);
    fill_bytes(b.w1q.ptr, static_cast<size_t>(n_local) * inter * (dim / 2), rng, false);
    fill_bytes(b.w1s.ptr, static_cast<size_t>(n_local) * inter * (dim / 32), rng, true);
    fill_bytes(b.w2q.ptr, static_cast<size_t>(n_local) * dim * (inter / 2), rng, false);
    fill_bytes(b.w2s.ptr, static_cast<size_t>(n_local) * dim * (inter / 32), rng, true);
    fill_bytes(b.w3q.ptr, static_cast<size_t>(n_local) * inter * (dim / 2), rng, false);
    fill_bytes(b.w3s.ptr, static_cast<size_t>(n_local) * inter * (dim / 32), rng, true);

    pocket::MoeMultiTokenFp4Workspace ws;
    ws.d_x_q = b.x_q.ptr;
    ws.d_x_scale = b.x_scale.ptr;
    ws.d_gate = b.gate.ptr;
    ws.d_up = b.up.ptr;
    ws.d_hidden_q = b.hidden_q.ptr;
    ws.d_hidden_scale = b.hidden_scale.ptr;
    ws.d_partials = b.partials.ptr;
    ws.tokens_cap = tokens;
    ws.pairs_cap = pairs;
    ws.dim = dim;
    ws.inter_dim = inter;
    if (!pocket::moe_multi_token_fp4_cuda_with_workspace(
            b.x.ptr, b.slot_expert.ptr, b.slot_starts.ptr, b.slot_tokens.ptr,
            b.pair_weights.ptr, b.w1q.ptr, b.w1s.ptr, b.w2q.ptr, b.w2s.ptr,
            b.w3q.ptr, b.w3s.ptr, b.y.ptr, tokens, slots, pairs, n_local,
            dim, inter, 7.0f, ws)) {
        throw std::runtime_error("multi-token launch failed");
    }
    check_cuda(cudaDeviceSynchronize(), "sync multi-token");

    for (int token = 0; token < tokens; ++token) {
        if (!pocket::moe_single_token_fp4_cuda(
                b.x.ptr + static_cast<size_t>(token) * dim,
                b.indices.ptr + static_cast<size_t>(token) * topk,
                b.weights.ptr + static_cast<size_t>(token) * topk,
                b.w1q.ptr, b.w1s.ptr, b.w2q.ptr, b.w2s.ptr, b.w3q.ptr, b.w3s.ptr,
                b.ref.ptr + static_cast<size_t>(token) * dim, topk, 0, n_local,
                dim, inter, 7.0f)) {
            throw std::runtime_error("single-token launch failed");
        }
    }
    check_cuda(cudaDeviceSynchronize(), "sync reference");
    std::vector<float> got(static_cast<size_t>(tokens) * dim);
    std::vector<float> want(got.size());
    check_cuda(cudaMemcpy(got.data(), b.y.ptr, got.size() * sizeof(float),
                          cudaMemcpyDeviceToHost), "copy multi output");
    check_cuda(cudaMemcpy(want.data(), b.ref.ptr, want.size() * sizeof(float),
                          cudaMemcpyDeviceToHost), "copy reference output");
    compare(got, want, rel_limit,
            "tokens=" + std::to_string(tokens) + " topk=" + std::to_string(topk) +
            " dim=" + std::to_string(dim));

    for (int iter = 0; iter < 3; ++iter) {
        if (!pocket::moe_multi_token_fp4_cuda_with_workspace(
                b.x.ptr, b.slot_expert.ptr, b.slot_starts.ptr, b.slot_tokens.ptr,
                b.pair_weights.ptr, b.w1q.ptr, b.w1s.ptr, b.w2q.ptr, b.w2s.ptr,
                b.w3q.ptr, b.w3s.ptr, b.y.ptr, tokens, slots, pairs, n_local,
                dim, inter, 7.0f, ws)) {
            throw std::runtime_error("repeat multi-token launch failed");
        }
        check_cuda(cudaDeviceSynchronize(), "sync repeated multi-token");
        std::vector<float> repeated(got.size());
        check_cuda(cudaMemcpy(repeated.data(), b.y.ptr, repeated.size() * sizeof(float),
                              cudaMemcpyDeviceToHost), "copy repeated output");
        if (std::memcmp(got.data(), repeated.data(), got.size() * sizeof(float)) != 0) {
            throw std::runtime_error("multi-token result is not bitwise reproducible");
        }
    }
}

void test_empty_routes() {
    constexpr int tokens = 3;
    constexpr int dim = 256;
    DeviceBuffer<float> x(static_cast<size_t>(tokens) * dim);
    DeviceBuffer<float> y(static_cast<size_t>(tokens) * dim);
    DeviceBuffer<int32_t> starts(1);
    DeviceBuffer<uint8_t> dummy(1);
    DeviceBuffer<int8_t> x_q(static_cast<size_t>(tokens) * dim);
    DeviceBuffer<float> x_scale(tokens);
    check_cuda(cudaMemset(starts.ptr, 0, sizeof(int32_t)), "zero starts");
    pocket::MoeMultiTokenFp4Workspace ws;
    ws.d_x_q = x_q.ptr;
    ws.d_x_scale = x_scale.ptr;
    ws.tokens_cap = tokens;
    ws.pairs_cap = 0;
    ws.dim = dim;
    ws.inter_dim = 128;
    if (!pocket::moe_multi_token_fp4_cuda_with_workspace(
            x.ptr, nullptr, starts.ptr, nullptr, nullptr, dummy.ptr, dummy.ptr,
            dummy.ptr, dummy.ptr, dummy.ptr, dummy.ptr, y.ptr, tokens, 0, 0, 8,
            dim, 128, 7.0f, ws)) {
        throw std::runtime_error("empty-route launch failed");
    }
    std::vector<float> host(static_cast<size_t>(tokens) * dim, 1.0f);
    check_cuda(cudaMemcpy(host.data(), y.ptr, host.size() * sizeof(float),
                          cudaMemcpyDeviceToHost), "copy empty output");
    if (std::any_of(host.begin(), host.end(), [](float v) { return v != 0.0f; })) {
        throw std::runtime_error("empty routes did not produce zero rows");
    }
}

}  // namespace

int main() {
    try {
        for (int tokens : {1, 2, 3, 6}) {
            run_case(tokens, 3, 8, 256, 128, 1000 + tokens, 4.0e-4f, false);
        }
        run_case(10, 8, 8, 256, 128, 77, 4.0e-3f, true);
        test_empty_routes();
        run_case(2, 6, 64, 4096, 2048, 4002, 8.0e-3f, false);
        std::cout << "[PASS] small-batch FP4 MoE\n";
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "[FAIL] " << ex.what() << "\n";
        return 1;
    }
}
